File: gherkin/dotnet/Gherkin/Pickles/Compiler.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Gherkin.CucumberMessages.Types;
using Gherkin.Events;
using Gherkin.Events.Args.Ast;

// ReSharper disable PossibleMultipleEnumeration

namespace Gherkin.Pickles
{
    public class Compiler
    {
        public List<Pickle> Compile(GherkinDocument gherkinDocument)
        {
            var pickles = new List<Pickle>();
            var feature = gherkinDocument.Feature;
            if (feature == null)
            {
                return pickles;
            }

            var language = feature.Language;
            var tags = feature.Tags;

            Build(pickles, language, tags, Enumerable.Empty<PickleStep>, feature.Children, gherkinDocument.Uri);

            return pickles;
        }

        protected virtual void Build(List<Pickle> pickles, string language, IEnumerable<Tag> tags,
            Func<IEnumerable<PickleStep>> parentBackgroundStepsFactory, IEnumerable<FeatureChild> children,
            string gherkinDocumentUri) 
        {
            if (children == null)
                return;

            Func<IEnumerable<PickleStep>> backgroundStepsFactory = parentBackgroundStepsFactory;
            foreach (var child in children)
            {
                if (child.Background != null)
                {
                    var previousFactory = backgroundStepsFactory;
                    backgroundStepsFactory = () =>
                        previousFactory().Concat(PickleSteps(child.Background));
                }
                else if (child.Rule != null)
                {
                    var mergedRuleTags = tags.Concat(child.Rule.Tags);
                    Build(pickles, language, mergedRuleTags, backgroundStepsFactory, child.Rule.Children, gherkinDocumentUri);
                }
                else if (child.Scenario != null)
                {
                    var scenario = child.Scenario;
                    if (!scenario.Examples.Any())
                    {
                        CompileScenario(pickles, backgroundStepsFactory, scenario, tags, language, gherkinDocumentUri);
                    }
                    else
                    {
                        CompileScenarioOutline(pickles, backgroundStepsFactory, scenario, tags, language, gherkinDocumentUri);
                    }
                }
            }
        }

        protected virtual void CompileScenario(List<Pickle> pickles,
            Func<IEnumerable<PickleStep>> backgroundStepsFactory, StepsContainer scenario, IEnumerable<Tag> featureTags,
            string language, string gherkinDocumentUri)
        {
            var steps = new List<PickleStep>();
            if (scenario.Steps.Any())
                steps.AddRange(backgroundStepsFactory());

            var scenarioTags = new List<Tag>();
            scenarioTags.AddRange(featureTags);
            scenarioTags.AddRange(scenario.Tags);

            steps.AddRange(PickleSteps(scenario));

            Pickle pickle = new Pickle(
                    IdGenerator.GetNextId(),
                    gherkinDocumentUri,
                    scenario.Name,
                    language,
                    steps,
                    PickleTags(scenarioTags),
                    new []{ scenario.Id }
            );
            pickles.Add(pickle);
        }

        protected virtual void CompileScenarioOutline(List<Pickle> pickles,
            Func<IEnumerable<PickleStep>> backgroundStepsFactory, StepsContainer scenarioOutline,
            IEnumerable<Tag> featureTags, string language, string gherkinDocumentUri)
        {
            foreach (var examples in scenarioOutline.Examples)
            {
                if (examples.TableHeader == null) continue;
                var variableCells = examples.TableHeader.Cells;
                foreach (var values in examples.TableBody)
                {
                    var valueCells = values.Cells;

                    var steps = new List<PickleStep>();
                    if (scenarioOutline.Steps.Any())
                        steps.AddRange(backgroundStepsFactory());

                    var tags = new List<Tag>();
                    tags.AddRange(featureTags);
                    tags.AddRange(scenarioOutline.Tags);
                    tags.AddRange(examples.Tags);

                    foreach(var scenarioOutlineStep in scenarioOutline.Steps)
                    {
                        string stepText = Interpolate(scenarioOutlineStep.Text, variableCells, valueCells);

                        PickleStep pickleStep = CreatePickleStep(
                                scenarioOutlineStep,
                                stepText,
                                CreatePickleArgument(scenarioOutlineStep, variableCells, valueCells),
                                new[] { scenarioOutlineStep.Id, values.Id }
                        );
                        steps.Add(pickleStep);
                    }

                    Pickle pickle = new Pickle(
                            IdGenerator.GetNextId(),
                            gherkinDocumentUri,
                            Interpolate(scenarioOutline.Name, variableCells, valueCells),
                            language, 
                            steps,
                            PickleTags(tags),
                            new[] { scenarioOutline.Id, values.Id }
                    );

                    pickles.Add(pickle);
                }
            }
        }

        protected virtual PickleStep CreatePickleStep(Step step, string text, PickleStepArgument argument, IEnumerable<string> astNodeIds)
        {
            return new PickleStep(IdGenerator.GetNextId(), text, argument, astNodeIds);
        }

        protected virtual PickleStepArgument CreatePickleArgument(Step argument)
        {
            var noCells = Enumerable.Empty<Cell>();
            return CreatePickleArgument(argument, noCells, noCells);
        }

        protected virtual PickleStepArgument CreatePickleArgument(Step step, IEnumerable<Cell> variableCells, IEnumerable<Cell> valueCells)
        {
            if (step.DataTable != null) {
                var t = step.DataTable;
                var rows = t.Rows;
                var newRows = new List<PickleRow>(rows.Count());
                foreach(var row in rows)
                {
                    var cells = row.Cells;
                    var newCells = new List<PickleCell>();
                    foreach(var cell in cells)
                    {
                        newCells.Add(
                                new PickleCell(
                                        Interpolate(cell.Value, variableCells, valueCells)
                                )
                        );
                    }
                    newRows.Add(new PickleRow(newCells));
                }
                return new PickleStepArgument
                    {
                        DataTable = new PickleTable(newRows)
                    };
            }

            if (step.DocString != null) {
                var ds = step.DocString;
                return
                    new PickleStepArgument
                    {
                        DocString = new PickleDocString(
                            Interpolate(ds.Content, variableCells, valueCells),
                            ds.MediaType == null ? null : Interpolate(ds.MediaType, variableCells, valueCells))
                    };
            } 
            
            return null;
        }

        protected virtual PickleStep[] PickleSteps(StepsContainer scenarioDefinition)
        {
            var result = new List<PickleStep>();
            foreach(var step in scenarioDefinition.Steps)
            {
                result.Add(PickleStep(step));
            }
            return result.ToArray();
        }

        protected virtual PickleStep PickleStep(Step step)
        {
            return CreatePickleStep(
                    step,
                    step.Text,
                    CreatePickleArgument(step),
                    new []{ step.Id }
            );
        }

        protected virtual string Interpolate(string name, IEnumerable<Cell> variableCells, IEnumerable<Cell> valueCells)
        {
            int col = 0;
            foreach (var variableCell in variableCells)
            {
                var valueCell = valueCells.ElementAt(col++);
                string header = variableCell.Value;
                string value = valueCell.Value;
                name = name.Replace("<" + header + ">", value);
            }
            return name;
        }

        protected virtual PickleLocation PickleLocation(Ast.Location location)
        {
            return new PickleLocation(location.Line, location.Column);
        }

        protected virtual List<PickleTag> PickleTags(List<Tag> tags)
        {
            var result = new List<PickleTag>();
            foreach(var tag in tags)
            {
                result.Add(PickleTag(tag));
            }
            return result;
        }

        protected virtual PickleTag PickleTag(Tag tag)
        {
            return new PickleTag(tag.Id, tag.Name);
        }
    }
}

File: gherkin/ruby/lib/gherkin/pickles/compiler.rb
module Gherkin
  module Pickles
    class Compiler
      def initialize(id_generator)
        @id_generator = id_generator
      end

      def compile(gherkin_document, source)
        pickles = []

        return pickles unless gherkin_document[:feature]
        feature = gherkin_document[:feature]
        language = feature[:language]
        tags = feature[:tags]

        compile_feature(pickles, language, tags, feature, source)
        pickles
      end

      private

      def compile_feature(pickles, language, tags, feature, source)
        feature_background_steps = []
        feature[:children].each do |child|
          if child[:background]
            feature_background_steps.concat(child[:background][:steps])
          elsif child[:rule]
            compile_rule(pickles, language, tags, feature_background_steps, child[:rule], source)
          else
            scenario = child[:scenario]
            if scenario[:examples].empty?
              compile_scenario(tags, feature_background_steps, scenario, language, pickles, source)
            else
              compile_scenario_outline(tags, feature_background_steps, scenario, language, pickles, source)
            end
          end
        end
      end

      def compile_rule(pickles, language, feature_tags, feature_background_steps, rule, source)
        tags = [].concat(feature_tags).concat(rule[:tags])

        rule_background_steps = feature_background_steps.dup
        rule[:children].each do |child|
          if child[:background]
            rule_background_steps.concat(child[:background][:steps])
          else
            scenario = child[:scenario]
            if scenario[:examples].empty?
              compile_scenario(tags, rule_background_steps, scenario, language, pickles, source)
            else
              compile_scenario_outline(tags, rule_background_steps, scenario, language, pickles, source)
            end
          end
        end
      end

      def compile_scenario(inherited_tags, background_steps, scenario, language, pickles, source)
        steps = scenario[:steps].empty? ? [] : [].concat(pickle_steps(background_steps))

        tags = [].concat(inherited_tags).concat(scenario[:tags])

        scenario[:steps].each do |step|
          steps.push(pickle_step(step))
        end

        pickle = {
          uri: source[:uri],
          id: @id_generator.new_id,
          tags: pickle_tags(tags),
          name: scenario[:name],
          language: language,
          astNodeIds: [scenario[:id]],
          steps: steps
        }
        pickles.push(pickle)
      end

      def compile_scenario_outline(inherited_tags, background_steps, scenario, language, pickles, source)
        scenario[:examples].reject { |examples| examples[:tableHeader].nil? }.each do |examples|
          variable_cells = examples[:tableHeader][:cells]
          examples[:tableBody].each do |values_row|
            value_cells = values_row[:cells]
            steps = scenario[:steps].empty? ? [] : [].concat(pickle_steps(background_steps))
            tags = [].concat(inherited_tags).concat(scenario[:tags]).concat(examples[:tags])

            scenario[:steps].each do |scenario_step|
              step = pickle_step_props(scenario_step, variable_cells, values_row)
              steps.push(step)
            end

            pickle = {
              uri: source[:uri],
              id: @id_generator.new_id,
              name: interpolate(scenario[:name], variable_cells, value_cells),
              language: language,
              steps: steps,
              tags: pickle_tags(tags),
              astNodeIds: [
                scenario[:id],
                values_row[:id]
              ],
            }
            pickles.push(pickle)

          end
        end
      end

      def interpolate(name, variable_cells, value_cells)
        variable_cells.each_with_index do |variable_cell, n|
          value_cell = value_cells[n]
          name = name.gsub('<' + variable_cell[:value] + '>', value_cell[:value])
        end
        name
      end

      def pickle_steps(steps)
        steps.map do |step|
          pickle_step(step)
        end
      end

      def pickle_step(step)
        pickle_step_props(step, [], nil)
      end

      def pickle_step_props(step, variable_cells, values_row)
        value_cells = values_row ? values_row[:cells] : []
        props = {
          id: @id_generator.new_id,
          astNodeIds: [step[:id]],
          text: interpolate(step[:text], variable_cells, value_cells),
        }
        if values_row
          props[:astNodeIds].push(values_row[:id])
        end

        if step[:dataTable]
          data_table = {
            dataTable: pickle_data_table(step[:dataTable], variable_cells, value_cells)
          }
          props[:argument] = data_table
        end
        if step[:docString]
          doc_string = {
            docString: pickle_doc_string(step[:docString], variable_cells, value_cells)
          }
          props[:argument] = doc_string
        end
        props
      end

      def pickle_data_table(data_table, variable_cells, value_cells)
        {
          rows: data_table[:rows].map do |row|
            {
              cells: row[:cells].map do |cell|
                {
                  value: interpolate(cell[:value], variable_cells, value_cells)
                }
              end
            }
          end
        }
      end

      def pickle_doc_string(doc_string, variable_cells, value_cells)
        props = {
          content: interpolate(doc_string[:content], variable_cells, value_cells)
        }
        if doc_string[:mediaType]
          props[:mediaType] = interpolate(doc_string[:mediaType], variable_cells, value_cells)
        end
        props
      end

      def pickle_tags(tags)
        tags.map {|tag| pickle_tag(tag)}
      end

      def pickle_tag(tag)
        {
          name: tag[:name],
          astNodeId: tag[:id]
        }
      end
    end
  end
end
